.. GeneRegulation documentation master file, created by
   sphinx-quickstart on Tue Mar 14 14:04:28 2017.
   You can adapt this file completely to your liking, but it should at least
   contain the root `toctree` directive.
   
   
France Genomique Workpackage 2.6 - Gene Regulation
==================================================

*Claire Rioualen - Sept. 19th, 2016*

This git repository holds shared code for the analysis of Next Generation Sequencing data related to gene regulation: ChIP-seq, RNA-seq, and related technologies.

The goals are multiple.

* Avoid duplication of efforts by sharing the developments of different bioinformaticians involved in ChIP-seq and RNA-seq projects.
* Ensure portability and re-usability of the code.
* Enable validation of the code by independent users.


We have chosen to use the Snakemake workflow management system[1] in order to build reproducible and flexible NGS analysis pipelines.
Snakemake workflows

Snakemake is a tool that enables the creation of analysis pipelines, based on the python language and the make concepts of rules and targets.

A rule contains the commands to generate a given target. A workflow can be defined as a series of files generated by successive rules.

Here, we present a repository of reusable NGS-specific rules, as well as a few workflow examples for standard ChIP-seq and RNA-seq analyses.

Welcome to GeneRegulation's documentation!
==========================================

Contents:

.. toctree::
   :maxdepth: 2



Indices and tables
==================

* :ref:`genindex`
* :ref:`modindex`
* :ref:`search`

