# Might be out of date

# Set parameters
if not "qsub" in config.keys():
    sys.exit("The parameter qsub must be defined in the config file")

if not "version" in config["genome"].keys():
    sys.exit("The parameter genome version must be defined in the config file")

rule bed_to_fasta:
    """Fetch sequences from UCSC to obtain a fasta file from a set of
    genomic coordinates described in a bed file. Requires RSAT installation. 
    Alternative is to use getfasta.rules .

    Example: 
    mkdir -p test/fetch_seq; cd test/fetch_seq; 
    wget http://pedagogix-tagc.univ-mrs.fr/rsat/demo_files/fetch-sequences_Schmidt_2011_mm9_CEBPA_SWEMBL_R0.12_702peaks.bed; 
    cd -
    snakemake --snakefile ${RSAT}/snakemake_files/chip-seq_motifs.py test/fetch_seq/fetch-sequences_Schmidt_2011_mm9_CEBPA_SWEMBL_R0.12_702peaks.fasta

    Required parameters:
        config["qsub"]

    Contributors: Jacques van Helden, Claire Rioualen
    """
    input: "{regions}.bed"
    output: "{regions}.fasta"
    params: 
        genome = config["genome"]["version"],
        qsub = config["qsub"] + " -e {regions}_fetchseq_qsub.err -o {regions}_fetchseq_qsub.out"
    log: "{regions}_fetchseq.log"
    benchmark: "{regions}_fetchseq_benchmark.txt"
    shell: "fetch-sequences -i {input} -header_format galaxy -genome {params.genome} -o {output} 2> {log}"
