# Check parameters
if not "qsub" in config.keys():
    sys.exit("The parameter qsub must be defined in the config file")

if not "genome" in config["dir"].keys():
    sys.exit("Genome dir must be defined in the 'dir' section of the config file")

if not "gtf_file" in config["genome"].keys():
    sys.exit("Genome dir must be defined in the 'dir' section of the config file")

if not "cufflinks" in config.keys():
    config["cufflinks"] = {}

if not "threads" in config["cufflinks"].keys():
    config["cufflinks"]["threads"] = "1"

if not "libtype" in config["cufflinks"].keys():
    config["cufflinks"]["libtype"] = "below"

# Rule
rule cufflinks:
    """Annotate a genome (infer transcript regions) based on RNA-seq data.
    Assemble transcripts from RNA-seq data and produce a file with
    the location of detected transcripts.

    Required parameters:
        config["qsub"]
        config["dir"]["genome"]
        config["genome"]["gtf_file"]

    Optional parameters:
        config["cufflinks"]["libtype"]
        config["cufflinks"]["threads"]

    Params to add:
        cuffmerge_dir
        assembly_files
        cuffmerge_transcripts

    Contributors: Jacques van Helden, Claire Rioualen
    """
    input:
        bam = "{reads}_sorted_pos.bam", \
        gtf = config["dir"]["genome"] + "/" + config["genome"]["gtf_file"]
    output:
        outdir="{reads}_cufflinks", \
        transcripts = "{reads}_cufflinks/transcripts.gtf"
    params: 
        threads = config["cufflinks"]["threads"], 
        libtype = config["cufflinks"]["libtype"], \
        qsub = config["qsub"] + " -e {reads}_cufflinks_qsub.err -o {reads}_cufflinks_qsub.out"
    log:  "{reads}_cufflinks.log"
    benchmark:  "{reads}_cufflinks_benchmark.txt"
    conda: "../envs/cufflinks.yaml"
    shell: "cufflinks --GTF-guide {input.gtf} --num-threads {params.threads}  --library-type {params.libtype}  --output-dir {output.outdir} {input.bam} 2> {log}" 

