# Set parameters
if not "qsub" in config.keys():
    sys.exit("The parameter qsub must be defined in the config file")

# Rule
rule genome_coverage_bigwig:    
    """Compute genome coverage from a bam-formatted file with aligned
    reads and produce a bigWig file. Uses bamCoverage tool from the
    deepTools suite. 

    Required parameters:
        config["qsub"]

    Contributors: Claire Rioualen
    """
    input: 
        bam = "{reads}.bam",
        bai = "{reads}.bam.bai"
    output: "{reads}.bw"
    params: 
#            output_prefix = "{reads}",
#            chromsize = re.sub('\.fa$', '', genome_file) + ".genome", 
            qsub = config["qsub"] + " -e {reads}_bam2bigwig_qsub.err -o {reads}_bam2bigwig_qsub.out"
    log:  "{reads}_bam2bigwig.log"
    benchmark:  "{reads}_bam2bigwig_benchmark.txt"
    shell: "bamCoverage -b {input.bam} -o {output} 2> {log}" 


