# Author Claire Rioualen 
# This is a draft rule for generating html reports


    
if not "title" in config.keys():
    sys.exit("Project title must be defined in the config file")

if not "dataset" in config.keys():
    sys.exit("Dataset ID must be defined in the config file")

if not "title" in config.keys():
    sys.exit("Project title must be defined in the config file")


#def html_report_inputs:
#(read_count, avg_read_length, gc_rate) = fastq_stats("/data/analyses/ChIP-seq_SE_GSE20870/fastq/GSM521934/GSM521934.fastq")

rule html_report:
    input: "synthesis.done"
    output: "{dir}/index.html"
    params:
        samples_ids = read_table(config["metadata"]["samples"])["ID"],
        samples_dir = config["dir"]["fastq"],
        trim_tools = config["trimming"],
        map_tools = config["mapping"],
        pc_tools = config["peakcalling"],
        outdir = "{dir}",
        dataset = config["dataset"],
        project_name = config["title"]
    log: "{dir}/index.log"
    benchmark: "{dir}/index_benchmark.tab"
    run:
#        "../../python_lib/generate_report.py"

## Draft python script to generate reports

# Usage: python generate_report.py path_to_configfile

import os
import sys
import pandas as pd
#import yaml
import collections

import web
import util
import statistics

dataset_id = snakemake.params["dataset"]
outputdir = snakemake.params["outdir"]
project_name = snakemake.params["project_name"]

# Framework/tools dictionary

#...

# Samples dictionary (separate ChIP and RNA?)

#samples_ids = snakemake.params["samples_ids"]
#samples_dir = snakemake.params["samples_dir"]
#trim_tools = snakemake.params["trim_tools"]
#map_tools = snakemake.params["map_tools"]
#pc_tools = snakemake.params["pc_tools"]


d_samples = {}
for sample in samples_ids:
    path = samples_dir + "/" + sample + "/" + sample + ".fastq"
    d_samples[sample] = {}
    d_samples[sample]["path"] = path

tfs = read_table(config["metadata"]["design"])['treatment'].split()
ctrl = read_table(config["metadata"]["design"])['control'].split()

compa = [m+"_vs_"+n for m,n in zip(tfs, ctrl)]

#tfs = ["Nac", "NtrC", "OmpR"]
#pc_tools = ["macs2", "spp", "homer", "swembl"]

peak_collections = []

d_peaks = {}
for tf in tfs:
    d_peaks[tf] = {}
    for pc in pc_tools:
#        key = tf + pc
        path = "results/peaks/" + tf + "_vs_" + "WT" + "/" + pc + "/" + tf + "_vs_" + "WT" + "_" + trim_tools + "_" + map_tools + "_" + pc + ".bed"
#        d_peaks[key] = {}
        d_peaks[tf][pc] = path

#print(d_peaks)

if __name__ == '__main__':
    web.indexPage(outputdir, project_name, dataset_id)
    web.homePage(outputdir, project_name, dataset_id, "home", description="")
    web.writeCategory(outputdir, project_name, dataset_id, "chip", "chip", d_samples, d_peaks)
    #web.downloadPage(outputdir, project_name, dataset_id, "download", description="")




### Read config.yml
##   analysis steps
##   filenames

#if not ("tools" in config.keys()):
#    sys.exit("The parameter config['tools'] should be specified in the config file.")

#step_list = []

## Check if there's a trimming step
#if ("trimming" in config["tools"].keys()):
#    trimming_step = True
#    step_list.append("trimming")
#    trimming_tools = config["tools"]["trimming"].split()

#else:
#    trimming_step=False

## Check if there's a mapping step
#if ("mapping" in config["tools"].keys()):
#    mapping_step = True
#    step_list.append("trimming")
#    mapping_tools = config["tools"]["mapping"].split()

#else:
#    mapping_step=False


## Check if there's a peak-calling step
#if ("peakcalling" in config["tools"].keys()):
#    peakcalling_step = True
#    step_list.append("trimming")
#    peakcalling_tools = config["tools"]["peakcalling"].split()

#else:
#    peakcalling_step = False




#if TRIMMING_TOOLS:
#    PREFIX = expand("{trimmer}_{aligner}", aligner=MAPPING_TOOLS, trimmer=TRIMMING_TOOLS)
#else:
#    PREFIX = expand("{aligner}", aligner=MAPPING_TOOLS)


### Read files -> script python

#    # If fastq files
#        # Generate Fastq stats
#        # Write to dictionary
#        # Copy files to reports dir

#    # If bam files
#        # Generate bam stats
#        # Write to dictionary
#        # Copy tdf files to reports dir

#    # If bed files
#        # Generate bed stats
#        # Write to dictionary
#        # Copy files to reports dir


## Write report



    # Write header

        # If fastq files
            # Write block

        # If bam files
            # Write block

        # If bed files
            # Write block

        # Plots (todo)

    # Write footer

## Write genome viewer session

    # IGV

    # Genome viewer


