# Author Claire Rioualen 
# This is a draft rule for generating IGV sessions


    
#if not "title" in config.keys():
#    sys.exit("Project title must be defined in the config file")

#if not "dataset" in config.keys():
#    sys.exit("Dataset ID must be defined in the config file")

#if not "title" in config.keys():
#    sys.exit("Project title must be defined in the config file")

#igv_session(filename, genome, gtf="", gff3="", peaks="", coverage="", transcripts_gtf=""):

# Genome & annotations
genome_dir = config["dir"]["genome"]


rule igv_session:
#    input: 
    output:  "{dir}/igv_session.xml"
    params:
        genome_fasta = os.path.join(genome_dir, config["genome"]["fasta_file"]),
        genome_gff3 = os.path.join(genome_dir, config["genome"]["gff3_file"]),
        genome_gtf = os.path.join(genome_dir, config["genome"]["gtf_file"]),
	genome_version = config["genome"]["version"]
#        samples_ids = read_table(config["metadata"]["samples"])["ID"],
#        samples_dir = config["dir"]["fastq"],
#        trim_tools = config["tools"]["trimming"],
#        outdir = "{dir}",
#        dataset = config["dataset"],
#        project_name = config["title"]
    log: "{dir}/index.log"
    benchmark: "{dir}/index_benchmark.tab"
    script:
        "../../python_lib/igv_session.py"


