# Check/set parameters
if not "qsub" in config.keys():
    sys.exit("The parameter qsub must be defined in the config file")

if not "peak-motifs" in config.keys():
    config["peak-motifs"] = {}

#if not "motif_db" in config["peak-motifs"].keys():
#    sys.exit("The parameter peak-motif:motif_db must be defined in the config file")

if not "motif_db" in config["peak-motifs"].keys():
    config["peak-motifs"]["motif_db"] = ""

if not "tasks" in config["peak-motifs"].keys():
    config["peak-motifs"]["tasks"] = "all"

if not "disco" in config["peak-motifs"].keys():
    config["peak-motifs"]["disco"] =  ""

# Rule
rule peak_motifs:
    """Motif discovery using the peak-motifs pipeline.

    Find out more @ http://floresta.eead.csic.es/rsat/help.peak-motifs.html

    Required parameters:
        config["qsub"]
        config["peak-motifs"]["motif_db"]

    Optional parameters:
        config["peak-motifs"]["tasks"]
        config["peak-motifs"]["disco"]

    Contributors: Claire Rioualen, Lucie Khamvongsa
    """
    input: "{caller_dir}/{peaks}.fasta"
    output: "{caller_dir}/peak-motifs/{peaks}_peak-motifs_synthesis.html"
    params:
        python_script = "SnakeChunks/scripts/python_lib/peak-motifs_soap.py",
        title = config["title"],
        bed_ctrl = "{caller_dir}/{peaks}.bed", \
        tasks = config["peak-motifs"]["tasks"], \
        disco = config["peak-motifs"]["disco"], \
        motif_db = config["peak-motifs"]["motif_db"], \
        outdir = "{caller_dir}/peak-motifs", \
        prefix="{peaks}_peak-motifs", \
        qsub = config["qsub"] + " -e {peaks}_qsub.err" \
                  + " -o {peaks}_qsub.out" 
    log: "{caller_dir}/{peaks}_peak-motifs.log"
    benchmark: "{caller_dir}/{peaks}_peak-motifs_bench.json"
    conda: "../envs/peakmotifs.yaml"
    shell:"""
export PEAKSEQ={input}

python {params.python_script} \
   -test ${{PEAKSEQ}} \
   -graph_title "{params.title}" \
   -disco dyads \
   -motif_db regulonDB \
   -server fungi
"""

#export PEAKSEQ=ChIP-seq/results/peaks/GSM1010220_vs_GSM1010224/macs2/GSM1010220_vs_GSM1010224_cutadapt_bowtie2_macs2.fasta

#python $RSAT/public_html//web_services/clients/python/peak-motifs_soap.py \
#   -test ${PEAKSEQ} \
#   -graph_title "FNR ChIP-seq;  Machin et al 201X" \
#   -disco dyads \
#   -motif_db regulonDB \
#   -server fungi



    #shell: "peak-motifs  -v 1 -i  {input} -disco {params.disco} -minol 5 -maxol 7 -motif_db {params.motif_db} -task {params.tasks} -prefix {params.prefix} -outdir {params.outdir} &> {log}"




